### PR TITLE
CDAP-2216: Program/App-Stream usage registry integration

### PR DESCRIPTION
**Pull request by jwang47**
https://issues.cask.co/browse/CDAP-2216
http://builds.cask.co/browse/CDAP-DUT1550-8

TODO:
- [ ] Test in distributed cluster. Some guice bindings may be missing.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java [hunk 1]
```diff
@@ -557,6 +562,12 @@ private void configureStreamInput(Job job, StreamBatchReadable stream) throws IO
 
     job.setInputFormatClass(StreamInputFormat.class);
 
+    try {
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java]: Is this the only way that a MR can read a stream? Isn't there a way to specify it via ``setInputDataset("stream:<name>"``?

**Comment by anew** [comment]
Overall looks good. I am just not sure whether this captures all possible ways of consuming a stream. See my comment.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java (hunk 1 above)
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java]: In that case too `configureStreamInput` is called - https://github.com/caskdata/cdap/blob/develop/cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java#L456

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/DefaultStreamWriter.java [hunk 2]
```diff
@@ -50,14 +54,27 @@
...
 
   @Inject
   public DefaultStreamWriter(@Assisted("namespaceId") String namespaceId,
+                             @Assisted("programId") Id.Program program,
+                             UsageRegistry usageRegistry,
                              DiscoveryServiceClient discoveryServiceClient) {
     this.namespaceId = namespaceId;
+    this.program = program;
     this.endpointStrategy = new RandomEndpointStrategy(discoveryServiceClient.discover(Constants.Service.STREAMS));
+    this.isStreamRegistered = Maps.newHashMap();
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/DefaultStreamWriter.java]: `write` can be called concurrently, right? If so we should make the map `ConcurrentMap`.

Comment (jwang47) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/DefaultStreamWriter.java]: I was assuming
1. `HashMap` won't fail if `write` is called concurrently
2. `ConcurrentMap` would incur more overhead, which would become significant

Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/DefaultStreamWriter.java]: Using data structures not designed for concurrency can lead to undefined behavior. I don't think using `ConcurrentMap` will add any significant overhead to stream operations.

Comment (jwang47) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/DefaultStreamWriter.java]: Fixed.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/DefaultStreamWriter.java [hunk 3]
```diff
@@ -50,14 +54,27 @@
...
 
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultStreamWriter.class);
+
   private final String namespaceId;
   private final EndpointStrategy endpointStrategy;
+  /**
+   * The program that is using this {@link StreamWriter}.
+   */
+  private final Id.Program program;
+  private final Map<Id.Stream, Boolean> isStreamRegistered;
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/DefaultStreamWriter.java]: Why a map to boolean value? Why not use a `Set`?

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/DefaultStreamWriter.java [hunk 4]
```diff
@@ -80,6 +97,15 @@ private URL getStreamURL(String stream, boolean batch) throws IOException {
   }
 
   private void writeToStream(Id.Stream stream, HttpRequest request) throws IOException {
+    if (!isStreamRegistered.containsKey(stream)) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/DefaultStreamWriter.java]: So there is race here if there are concurrent calls to `writeToStream` (both threads `containsKey` return false, they'll both register and both update the map). I guess it's ok to make more than one register call, but strictly speaking it's promoting a wrong assumption here. If you want to make it atomic, you should use `putIfAbsent`

Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/DefaultStreamWriter.java]: Strictly speaking, you'll need a check->sync->check->update logic.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/StreamWriterFactory.java [hunk 5]
```diff
@@ -25,8 +26,9 @@
 public interface StreamWriterFactory {
   /**
    * @param namespaceId the namespaceId for which to return a {@link StreamWriter}
+   * @param programId the programId that will be using the {@link StreamWriter}
    * @return a {@link StreamWriter} for the specified namespaceId
    */
-  StreamWriter create(@Assisted("namespaceId") String namespaceId);
+  StreamWriter create(@Assisted("namespaceId") String namespaceId, @Assisted("programId") Id.Program programId);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/StreamWriterFactory.java]: Isn't Id.Program contains namespaceId? Why pass it twice?

**Comment by anew** [comment]
@jwang47 please do not squash all your commits into a single commit until the review is done. Here you only needed a few lines of change to address comments, but reviewers now have to read 500 lines of changes again. It is much easier to do an incremental review if the comments are addressed in a separate commit.

**Comment by jwang47** [comment]
Opened new PR for unaddressed comments: https://github.com/caskdata/cdap/pull/2205

**Pull request merged**